File: backend/src/modules/storage/storage.service.ts
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  S3Client,
  PutObjectCommand,
  DeleteObjectCommand,
  CreateBucketCommand,
  HeadBucketCommand,
  PutBucketPolicyCommand,
} from '@aws-sdk/client-s3';
import { v4 as uuidv4 } from 'uuid';
import * as path from 'path';

@Injectable()
export class StorageService implements OnModuleInit {
  private readonly logger = new Logger(StorageService.name);
  private s3: S3Client;
  private bucket: string;
  private publicUrl: string;

  constructor(private configService: ConfigService) {
    const endpoint = this.configService.get<string>('MINIO_ENDPOINT', 'http://127.0.0.1');
    const port = this.configService.get<number>('MINIO_PORT', 9000);
    const accessKey = this.configService.get<string>('MINIO_ACCESS_KEY', 'minioadmin');
    const secretKey = this.configService.get<string>('MINIO_SECRET_KEY', 'minioadmin');

    this.bucket = this.configService.get<string>('MINIO_BUCKET', 'products');
    this.publicUrl = this.configService.get<string>('MINIO_PUBLIC_URL', 'http://127.0.0.1:9000');

    this.s3 = new S3Client({
      endpoint: `${endpoint}:${port}`,
      region: 'us-east-1', // MinIO requires any region string
      forcePathStyle: true, // CRITICAL for MinIO — prevents AWS virtual-host URL style
      credentials: {
        accessKeyId: accessKey,
        secretAccessKey: secretKey,
      },
    });
  }

  // Auto-create bucket on startup if it doesn't already exist
  async onModuleInit() {
    await this.ensureBucketExists(this.bucket);
  }

  private async ensureBucketExists(bucketName: string) {
    try {
      // Check if bucket already exists
      await this.s3.send(new HeadBucketCommand({ Bucket: bucketName }));
      this.logger.log(`Bucket "${bucketName}" already exists.`);
    } catch {
      // Bucket doesn't exist — create it
      try {
        await this.s3.send(new CreateBucketCommand({ Bucket: bucketName }));
        this.logger.log(`Bucket "${bucketName}" created.`);

        // Set bucket to public-read so frontend can render images directly
        const publicPolicy = {
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: '*',
              Action: 's3:GetObject',
              Resource: `arn:aws:s3:::${bucketName}/*`,
            },
          ],
        };
        await this.s3.send(
          new PutBucketPolicyCommand({
            Bucket: bucketName,
            Policy: JSON.stringify(publicPolicy),
          }),
        );
        this.logger.log(`Bucket "${bucketName}" set to public-read.`);
      } catch (err) {
        this.logger.error(`Failed to create bucket "${bucketName}":`, err);
      }
    }
  }

  /**
   * Upload a file to MinIO.
   * @param file - Multer file object
   * @param folder - Optional subfolder prefix e.g. 'products' or 'categories'
   * @returns Full public URL of the uploaded image
   */
  async uploadFile(file: Express.Multer.File, folder = 'products'): Promise<string> {
    const ext = path.extname(file.originalname);
    const key = `${folder}/${uuidv4()}${ext}`;

    await this.s3.send(
      new PutObjectCommand({
        Bucket: this.bucket,
        Key: key,
        Body: file.buffer,
        ContentType: file.mimetype,
      }),
    );

    return `${this.publicUrl}/${this.bucket}/${key}`;
  }

  /**
   * Delete a file from MinIO by its full URL.
   * @param fileUrl - Full public URL of the file to delete
   */
  async deleteFile(fileUrl: string): Promise<void> {
    try {
      // Extract the key from the full URL
      const urlPath = new URL(fileUrl).pathname;
      // pathname looks like /products/products/uuid.jpg
      // we need to strip the leading /bucket-name/ to get the key
      const key = urlPath.replace(`/${this.bucket}/`, '');

      await this.s3.send(
        new DeleteObjectCommand({
          Bucket: this.bucket,
          Key: key,
        }),
      );
      this.logger.log(`Deleted file: ${key}`);
    } catch (err) {
      this.logger.warn(`Could not delete file ${fileUrl}:`, err);
    }
  }
}

File: ecommerce/src/main/java/com/ecommerce/service/StorageService.java
package com.ecommerce.service;

import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.*;

import java.io.IOException;
import java.net.URI;
import java.util.UUID;

@Service
@Slf4j
public class StorageService {

    private S3Client s3;

    @Value("${minio.endpoint}")
    private String endpoint;

    @Value("${minio.port}")
    private int port;

    @Value("${minio.access-key}")
    private String accessKey;

    @Value("${minio.secret-key}")
    private String secretKey;

    @Value("${minio.bucket}")
    private String bucket;

    @Value("${minio.public-url}")
    private String publicUrl;

    @PostConstruct
    public void init() {
        this.s3 = S3Client.builder()
                .endpointOverride(URI.create(endpoint + ":" + port))
                .region(Region.US_EAST_1)
                .forcePathStyle(true)
                .credentialsProvider(StaticCredentialsProvider.create(
                        AwsBasicCredentials.create(accessKey, secretKey)))
                .build();

        ensureBucketExists();
    }

    private void ensureBucketExists() {
        try {
            s3.headBucket(HeadBucketRequest.builder().bucket(bucket).build());
            log.info("Bucket \"{}\" already exists.", bucket);
        } catch (S3Exception e) {
            try {
                s3.createBucket(CreateBucketRequest.builder().bucket(bucket).build());
                log.info("Bucket \"{}\" created.", bucket);
            } catch (Exception ex) {
                log.error("Failed to create bucket \"{}\": {}", bucket, ex.getMessage());
            }
        }

        // Always try to apply the public-read policy just in case it doesn't have it
        try {
            String publicPolicy = """
                    {
                      "Version": "2012-10-17",
                      "Statement": [
                        {
                          "Effect": "Allow",
                          "Principal": "*",
                          "Action": "s3:GetObject",
                          "Resource": "arn:aws:s3:::%s/*"
                        }
                      ]
                    }
                    """.formatted(bucket);

            s3.putBucketPolicy(PutBucketPolicyRequest.builder()
                    .bucket(bucket)
                    .policy(publicPolicy)
                    .build());
            log.info("Bucket \"{}\" set to public-read.", bucket);
        } catch (Exception ex) {
            log.warn("Warning: Could not set bucket policy for \"{}\": {}", bucket, ex.getMessage());
        }
    }

    /**
     * Upload a file to MinIO.
     * 
     * @param file   Multipart file from the request
     * @param folder Subfolder prefix e.g. "products" or "categories"
     * @return Full public URL of the uploaded image
     */
    public String uploadFile(MultipartFile file, String folder) throws IOException {
        String originalFilename = file.getOriginalFilename();
        String ext = "";
        if (originalFilename != null && originalFilename.contains(".")) {
            ext = originalFilename.substring(originalFilename.lastIndexOf("."));
        }

        String key = folder + "/" + UUID.randomUUID() + ext;

        s3.putObject(
                PutObjectRequest.builder()
                        .bucket(bucket)
                        .key(key)
                        .contentType(file.getContentType())
                        .build(),
                RequestBody.fromBytes(file.getBytes()));

        return publicUrl + "/" + bucket + "/" + key;
    }

    /**
     * Delete a file from MinIO by its full URL.
     */
    public void deleteFile(String fileUrl) {
        try {
            URI uri = URI.create(fileUrl);
            String path = uri.getPath();
            // path looks like /products/products/uuid.jpg — strip leading /bucket/
            String key = path.replaceFirst("/" + bucket + "/", "");

            s3.deleteObject(DeleteObjectRequest.builder()
                    .bucket(bucket)
                    .key(key)
                    .build());
            log.info("Deleted file: {}", key);
        } catch (Exception e) {
            log.warn("Could not delete file {}: {}", fileUrl, e.getMessage());
        }
    }
}
